src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -383,67 +374,68 @@ void ShenandoahOldGeneration::transition_to(State new_state) {
+//               +------------> |     FILLING     |
+//               |   +--------> |                 |
+//               |   |          +-----------------+
+//               |   |            |
+//               |   |            | Filling Complete
+//               |   |            v
+//               |   |          +-----------------+
+//               |   +--------> |     WAITING     |
+//               |   |          |  FOR BOOTSTRAP  |
+//               |   |          +-----------------+
+//               |   |            |
+//               |   |            | Reset Bitmap
+//               |   |            v
+//               |   |          +-----------------+     +----------------------+
+//               |   |          |    BOOTSTRAP    | <-> |       YOUNG GC       |
+//               |   |          |                 |     | (RSet Parses Region) |
+//               |   |          +-----------------+     +----------------------+
+//               |   |            |
+//               |   |            | Old Marking
+//               |   |            v
+//               |   |          +-----------------+     +----------------------+
+//               |   |          |     MARKING     | <-> |       YOUNG GC       |
+//               |   +--------- |                 |     | (RSet Parses Region) |
+//               |              +-----------------+     +----------------------+
+//               |                |
+//               |                | Has Evacuation Candidates
+//               |                v
+//               |              +-----------------+     +--------------------+
+//               |              |    EVACUATING   | <-> |      YOUNG GC      |
+//               +------------- |                 |     | (RSet Uses Bitmap) |
+//                              +-----------------+     +--------------------+
+//
+//
 //
 void ShenandoahOldGeneration::validate_transition(State new_state) {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   switch (new_state) {
-    case IDLE:
-      // GC cancellation can send us back to IDLE from any state.
-      assert(!heap->is_concurrent_old_mark_in_progress(), "Cannot become idle during old mark.");
-      assert(_old_heuristics->unprocessed_old_collection_candidates() == 0, "Cannot become idle with collection candidates");
-      assert(!heap->is_prepare_for_old_mark_in_progress(), "Cannot become idle while making old generation parseable.");
-      assert(heap->young_generation()->old_gen_task_queues() == nullptr, "Cannot become idle when setup for bootstrapping.");
-      break;
     case FILLING:
-      assert(_state == IDLE || _state == WAITING_FOR_FILL, "Cannot begin filling without first completing evacuations, state is '%s'", state_name(_state));
-      assert(heap->is_prepare_for_old_mark_in_progress(), "Should be preparing for old mark now.");
+      assert(heap->is_old_bitmap_stable(), "Cannot begin filling without first completing marking, state is '%s'", state_name(_state));
```

Comment:
Do you also want to say:

```
assert(_state > FILLING && _state != WAITING_FOR_BOOTSTRAP, "Illegal state transition"):
```

Reply:
No - at the end of a _GLOBAL_ collection, it may put the old generation in this state without progressing through the other states.